### PR TITLE
Initial concept mapping update

### PR DESCRIPTION
**Pull request by ViktorDomazetoski**
# Summary 
This is a first version of the concept mapping update. The reason why I say "first" is that there are several issues that need to be fixed before we finalize the concept mapping, but I still thought it would be useful to have someone review the logic. 

# How it works
1. The `initialize_benthic_concepts()` function initializes the benthic attribute to label mapping as a dataframe using the class_to_concepts file (*this is currently kept in a temporary location in the S3 bucket*). 
2. If given the `global_id2source` (which is just acquired from the Registry), the mapping is subset to keep only the source_label_class_names which are within the registry. (*this currently doesn't really work well, because of discrepancies in the naming within the class_to_concepts file and the dataset annotations*). This subsetting is done so that we don't include any concept values that aren't found in our dataset (e.g. if we don't have any plants in the globalid2source, the value taxonomic kingdom==plantae doesn't make sense to add to our concepts), as well as concepts that have no diversity (e.g. if we don't have any massive corals, we won't have any massive==True concepts, so it is pointless to evaluate on that concept when it might only be False or not given).
3. The taxonomic concepts are generated based on the above mapping subset in the `initialize_taxonomic_concept_mapping()` function. The unique values of each column (taxonomic rank) are mapped using `generate_taxonomic_id_mapping()` using the following logic: 1) the value "not_given" is always mapped to 0 (so as to not be included in the loss function calculations); 2) the value "none" (if exists) is mapped to 1; 3)the rest of the values are mapped starting from 2, sorted alphabetically. 
This returns a torch.Tensor lookup table (Nx7, where N is the length of global_id2source, and the columns represent 7 taxonomic ranks from kingdom to species) and a dict[str, dict] containing the name2id mapping for each taxonomic rank.
4. The one hot concepts (which are technically not one hot since they can be not_given, false and true), are generated based on the above mapping subset in the `initialize_one_hot_concept_mapping()` function. The unique values of each column (which includes morphologic, health and noncoral concepts) are mapped using `generate_one_hot_id_mapping()` using the following logic: the value "not_given" is always mapped to 0 (so as to not be included in the loss function calculations); If the concept contains both true and false values, then map True to 2 and False to 1; Otherwise, map all values to 0 since they don't represent a meaningful concept.. This returns a torch.Tensor lookup table (Nx33, where N is the length of global_id2source, and the columns represent 24 morphologic concepts, 2 health concepts, 7 non-coral concepts) and a dict[str, dict] containing the name2id mapping for each taxonomic rank.
5. The taxonomic concept lookup table and one hot concept look up table are concatenated to create the final source_label_to_concept look up table (Nx40) 
6. The function returns `df_mapping` (which is a subset of the class_to_concepts DataFrame based on the global_id2source classes), `source_to_concepts` (which is the Nx40 lookup table), `taxonomic_mapping_dictionary` (which contains the name2id of the taxonomic concepts), `one_hot_mapping_dictionary` (which contains the name2id for the one-hot encoded concepts. 

# Problems & Questions
- We need to specify a permanent space to store files such as the class_to_concepts.csv among other label mappings and dataframes. 
- The class names in Mermaid are capitalized, while the ones in the class_to_concepts.csv are all lowercase.
- CoralNet labels are kept as and read as IDs in the annotations file, and are used as such in the LabelMapping API, but in the concept mapping they are the full names. IMO we should update everything to use the full label names (rather than using the IDs). 
- The class_to_concepts.csv contains FALSE and False, and TRUE and True. Need to standardize

### CODE REVIEW

**Comment by josauder** [comment]
> We need to specify a permanent space to store files such as the class_to_concepts.csv among other label mappings and dataframes.

I think we should put it on git into configs/

> The class names in Mermaid are capitalized, while the ones in the class_to_concepts.csv are all lowercase.

IMO we should make everything to lowercase everywhere we can.

> CoralNet labels are kept as and read as IDs in the annotations file, and are used as such in the LabelMapping API, but in the concept mapping they are the full names. IMO we should update everything to use the full label names (rather than using the IDs).

Yes, I think class names are good also for debugging/clarity etc. Do you want to do it as part of this PR?

> The class_to_concepts.csv contains FALSE and False, and TRUE and True. Need to standardize

Yes, let's use `True` and `False` or even `0` and `1`.

File: mermaidseg/dataset_reconciliation/concepts.py [hunk 1]
```diff
@@ -81,38 +81,195 @@ def generate_hierarchy_path(label: str, hierarchy_dict: dict[str, str]) -> list[
...
+        (df_mapping_taxonomic.shape[0] + 1, len(taxonomic_concept_columns)), dtype=np.int64
+    )
+    source_to_taxonomic_concepts_np[df_mapping_taxonomic["global_id"].to_numpy()] = (
+        df_mapping_taxonomic[taxonomic_concept_columns].to_numpy()
+    )
+    source_to_taxonomic_concepts = torch.from_numpy(source_to_taxonomic_concepts_np).long()
+    return source_to_taxonomic_concepts, taxonomic_name2id
+
+
+def initialize_one_hot_concept_mapping(
```
Comment (josauder) [inline mermaidseg/dataset_reconciliation/concepts.py]: You mean multi-hot, right? Multiple taxonomic concepts can be 1, and we'll need a channel-wise cross-entropy loss instead of a softmax-crossentropy.

The taxonomic-concepts are 1-hot per hierarchy

**Comment by yeelauren** [comment]
Dup of #98
